app/src/org/commcare/android/nfc/NfcActivity.java

```diff
@@ -25,12 +29,17 @@
 
     protected static final String NFC_PAYLOAD_MULT_TYPES_ARG = "types";
     protected static final String NFC_PAYLOAD_SINGLE_TYPE_ARG = "type";
+    protected static final String NFC_ENCRYPTION_KEY_ARG = "encryption_key";
     protected static final String NFC_DOMAIN_ARG = "domain";
+    protected static final String NFC_ENCRYPTION_SCHEME = "encryption_aes_v1";
```

Comment:
@ctsims This is what I am adding to the encrypted message before writing to the tag. I didn't think case_id would always be available so decided to not incorporate it in this. Does this look ok ?

Reply:
As small point of personal preference I'd probably think of this scheme tag not as an encryption scheme, but as a general serialization / deserialization scheme tag.

Wasn't the `case_id` thing a pretty hard part of the requirements that they really wanted in place? I think it should be possible to handle both the case where a key is available and one where it isn't. We could have `scheme_encryption_aes_v1` and `scheme_encryption_and_key_aes_v1` and fork behavior on the latter to have a delimiter encoded tag or something.

Alternatively, don't the tags themselves come with keys or buckets that we're writing to? Maybe those can be how we determine where things go

Reply:
>  but as a general serialization / deserialization scheme tag.

I am not sure I follow what you mean by a serialization tag. Would you be able to point me to one. 

> Wasn't the case_id thing a pretty hard part of the requirements that they really wanted in place?..

You mean add another extra to the NFC callout called "case_id" and use "scheme_encryption_and_($case_id)_aes_v1" as a scheme instead ? I was looking for a generalized key for all projects but agree we can use `case_id` if available. OOC, the benefit this will have is make the tag we are adding more unique and dynamic ?

Reply:
On your first point: Oh, sorry, I just meant the title of the field. IE: This is 'tagging' the encryption scheme, but I think its purpose is more of a general "scheme" of what we are writing, not just how we are encrypting it. IE: `NFC_ENCRYPTION_SCHEME -> NFC_PAYLOAD_SCHEME`

Reply:
On your second point:

I wasn't sure, but I thought the NFC code wrote to buckets, so I was hoping we could just write to a bucket based on the case Id, but that might not be correct.

If we can't write to a named bucket, I wasn't quite thinking what you wrote, I was thinking more that we would write one of three things to the card. I'll refer to the `case_id` as `entity_id` below, since I don't think it's necessarily specific to the case id guid.

```
[No key, No entity_id] - > $RAW_PAYLOAD
[key, no entity_id] -> scheme_aes_v1|$RAW_PAYLOAD
[key, entity_id] -> scheme_entity_aes_v1|$ENTITY_ID|$RAW_PAYLOAD
```

Alternatively if you wanted a universal scheme, I think if we had delimiting, you could just leave the entity id blank unless it was present.

```
[No key, No entity_id] - > $RAW_PAYLOAD
[key, no entity_id] -> scheme_aes_v1||$RAW_PAYLOAD
[key, entity_id] -> scheme_aes_v1|$ENTITY_ID|$RAW_PAYLOAD
```

We'd need a delimiter that we had confidence couldn't be a part of the entity id (Could just throw an exception if it's in there) or a part of the base64 character set.

Reply:
@ctsims Added this in https://github.com/dimagi/commcare-android/pull/2502/commits/1a299ea0457ee729c23800cc04108bfac01ed4ca